iree/turbine/kernel/wave/templates/gqa_vanilla_attention.py

```diff
@@ -0,0 +1,221 @@
+# Copyright 2025 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import iree.turbine.kernel.lang as tkl
+import iree.turbine.kernel.wave as tkw
+from iree.turbine.kernel.lang.global_symbols import *
+from iree.turbine.kernel.wave.constraints import MMAType
+from iree.turbine.kernel.wave.utils.general_utils import (
+    torch_dtype_to_wave,
+)
+from .attention_common import *
+import math
+import torch
+from typing import Optional
+
+
+def get_gqa_bshd_attention_kernel(
+    shape: AttentionShape,
+    mfma_variant: tuple[MMAType, MMAType],
+    input_dtype: Optional[torch.dtype] = torch.float16,
+    output_dtype: Optional[torch.dtype] = torch.float32,
+    is_causal: Optional[bool] = False,
+    layer_scaling: Optional[float] = None,
+    sliding_window_size: Optional[int] = -1,
+    q_scale: Optional[float] = 1.0,
+    k_scale: Optional[float] = 1.0,
+    v_scale: Optional[float] = 1.0,
+    use_fp8: Optional[bool] = False,
+):
+
+    if sliding_window_size > 0 and not is_causal:
+        raise NotImplementedError(
+            "Sliding window is only supported for causal attention."
+        )
+
+    # Determine dtype of operands.
+    wave_input_dtype = torch_dtype_to_wave(input_dtype)
+    wave_output_dtype = torch_dtype_to_wave(output_dtype)
+
+    LOG2E = 1.44269504089
+    dk_sqrt = math.sqrt(1.0 / shape.head_size)
+    layer_scaling = (layer_scaling or dk_sqrt) * LOG2E
+
+    f8_dtype: torch.dtype = torch.float8_e4m3fnuz
+    F8_DTYPE = torch_dtype_to_wave(f8_dtype)
+    F8_MAX = torch.finfo(f8_dtype).max
+
+    # maximum expected value from attention softmax
+    ATTENTION_SOFTMAX_MAX = 1.0
+
+    # FP8 offset
+    # If we need to truncate to fp8 post softmax we apply a scaling to use the
+    # full fp8 range. We can do this with a offset as post `exp2` this equates
+    # to multiplying by a static value. We are able to do this as `max` and
+    # `sum` are scaled by the same value so the end result is the same.
+    FP8_OFFSET_VAL = math.log2(F8_MAX / ATTENTION_SOFTMAX_MAX)
+
+    # Dequant Tensor Scaling
+    DEQUANT_QK = q_scale * k_scale
+    if use_fp8:
```

Comment:
A TODO for coming PRs is to guard more stuff with the `use_fp8` flag, I reckon we don't to apply FP8_OFFSET_VAL when it's no fp8, as well as the v_scale(i.e someone can theoretically turn on v_scale but not fp8).

Reply:
good point, will fix.

Reply:
I am surprised that this passed the fp16 tests.